### PR TITLE
Annotation usage simplification

### PR DESCRIPTION
**Pull request by Vaspra**
-- Collection --
- Grammar/docstring/format updates.
- Proper caching for lookups (v1 and v2).
- Utility functions in Collection to hide private methods (particularly around using lookups to swap from image/row spaces).
- Enhanced checking and exception raising.
- Multi-threaded save/download actions now check for exceptions in worker threads.

-- Annotation --
- Grammar/docstring/format updates.
- Moved reliance on internal data to getters to reduce chance of errors.
- Removed unnecessary constructor in AnnotationMask.
- Moved boolmask checking into dedicated static method, with a choice to parse as [h, w, N] or [h, w] arrays.

### CODE REVIEW

File: zegami_sdk/annotation.py [hunk 1]
```diff
@@ -18,9 +19,8 @@ class _Annotation():
     TYPE = None
     UPLOADABLE_DESCRIPTION = None
 
-    def __init__(self, collection, annotation_data, source=None):
-        """
-        !! STOP !! Instantiate a non-hidden subclass instead.
+    def __init__(self, collection, id, row_index, source=0):
+        """!! STOP !! Instantiate a non-hidden subclass instead.
```
Comment (ahaith) [inline zegami_sdk/annotation.py:23]: Check out AbstractBaseClass which makes it impossible to instantiate the base class. https://docs.python.org/3/library/abc.html

File: zegami_sdk/annotation.py [hunk 2]
```diff
@@ -30,8 +30,10 @@ def __init__(self, collection, annotation_data, source=None):
         """
 
         self._collection = collection  # Collection instance
-        self._source = source  # Source instance
-        self._data = annotation_data  # { imageset_id, image_index, type, annotation }
+        self._imageset_id = collection._get_imageset_id(source=source)
+        self._row_index = row_index
+
+        self._data = collection.fetch_annotation_data(self._row_idx)  # { imageset_id, image_index, type, annotation }
```
Comment (ahaith) [inline zegami_sdk/annotation.py:36]: As things stand, annotations apply to the specific image (i.e. source) So the same row could have an annotation for the image of one source, but not for the other. The same annotation could even be associated to multiple rows if multiple rows refer to the same image.
There are clearly scenarios where it'd make more sense for the annotation to transcend the sources, but also cases where it doesn't.

Not sure how best to deal with that, perhaps include source in `fetch_annotation_data`?

Comment (ahaith) [reply zegami_sdk/annotation.py:36]: In fact, I don't see `fetch_annotation_data` in collection.py. Is there a change missing from this PR?

File: zegami_sdk/collection.py [hunk 3]
```diff
@@ -305,28 +390,38 @@ def get_image_urls(self, rows, source=0, generate_signed_urls=False,
             return signed_route_urls
 
     def download_annotation(self, annotation_id):
-        """
-        Converts an annotation_id into downloaded annotation data.
+        """Converts an annotation_id into downloaded annotation data.
```
Comment (ahaith) [inline zegami_sdk/collection.py:393]: Multiline docstrings in our other python repos have the opening `"""` on its own line (as per before this change).
I don't have a strong opinion about this, but it does seem better to have a consistent style.

**Review by ahaith — COMMENTED** [review]
Looks fine although in future better to do these kinds of changes as a set of small self-contained PRs, even if they're just little ad-hoc fixups.

There does seem to be a missing function here so please look into that - would be easier to notice that if this PR was *just* changes to the annotation stuff!

File: zegami_sdk/annotation.py (hunk 1 above)
Comment (Vaspra) [reply zegami_sdk/annotation.py:23]: Solved in new commit, doesn't actually require abc as this functionality is enforced a different way

File: zegami_sdk/annotation.py (hunk 2 above)
Comment (Vaspra) [reply zegami_sdk/annotation.py:36]: Not sure how this creeped in, my version doesn't even use this any more! New commit fixes

File: zegami_sdk/collection.py (hunk 3 above)
Comment (Vaspra) [reply zegami_sdk/collection.py:393]: All of collection.py and annotation.py fixed up to suit this

**Review by ahaith — APPROVED** [review]

**Pull request merged**